application/cms/models.py

```diff
@@ -102,18 +102,84 @@ def process(value):
 
 
 class FrequencyOfRelease(db.Model):
+    __tablename__ = "frequency_of_release"
+
     id = db.Column(db.Integer, primary_key=True)
     description = db.Column(db.String(), nullable=False)
     position = db.Column(db.Integer, nullable=False)
 
 
 class TypeOfStatistic(db.Model):
+    __tablename__ = "type_of_statistic"
+
     id = db.Column(db.Integer, primary_key=True)
     internal = db.Column(db.String(), nullable=False)
     external = db.Column(db.String(), nullable=False)
     position = db.Column(db.Integer, nullable=False)
 
 
+class DataSource(db.Model):
+    __tablename__ = "data_source"
+
+    # columns
+    id = db.Column(db.Integer, primary_key=True)
+    title = db.Column(db.String(255), nullable=True)
+
+    type_of_data = db.Column(ArrayOfEnum(db.Enum(TypeOfData, name="type_of_data_types")), default=[], nullable=True)
+    type_of_statistic_id = db.Column(db.Integer, nullable=True)
+
+    publisher_id = db.Column(db.String(255), nullable=True)
+    external_url = db.Column(db.String(255), nullable=True)
+    publication_date = db.Column(db.String(255), nullable=True)
+    note_on_corrections_or_updates = db.Column(db.TEXT, nullable=True)
+
+    frequency_of_release_id = db.Column(db.Integer, nullable=True)
+    frequency_of_release_other = db.Column(db.TEXT, nullable=True)
```

Comment:
Feels like this could be `String(255)`  too? The longest we currently have is `Every 7 years (further publications dependent on further surveys being commissioned)` so the copying over migration would be fine.

Reply:
@TheDoubleK @samuelhwilliams I'm not convinced that we should have any database character limits on strings which the user enters through the interface. There's no storage benefit, and we should be imposing any constraints as validation errors in the WTForms model instead, so that the user gets a nice error message rather than a 500 error. We _could_ duplicate these constraints in the database as well as the application code, but I’m not sure there's much benefit to that, and it means we have to maintain the same arbitrary number in two places (which are bound to go out-of-sync in future).

Reply:
Yep, sensible, done. 👍

Reply:
@frankieroberto If we want to enforce length constraints, the database is _the_ place to do this safely. It's our last chance, and best opportunity, to enforce the integrity of our data before we store it forever. We can still manage validation on the frontend, either via duplication (preferred) or catching integrity errors from the database. Duplication here seems a reasonable trade-off for the security added by database constraints.

Reply:
@samuelhwilliams I'm still not convinced. If you have validation at the application level, then the database constraints are only helpful if you have an error in the application. And I'm not sure that's a worthwhile trade-off against maintaining the duplication of the constraint.

It also feels a bit weird to me to enforce character length validations in the database, but not other validations, like URL strings being valid, or dates being in a given range, or whatever.

I've also been bitten in the past by Postgres being really slow to alter column constraints (eg changing `varchar(255)` to `text`, or increasing the varchar limit), as it basically has to lock the table for writes, then copy all the data to a new table, then delete the old one. Which can take a while if the table is big.

It’s not black and white - eg generally I find foreign keys and `not null` constraints to be helpful – but you have to draw the line at putting business logic into the database somewhere...

Reply:
I guess I fundamentally disagree. I think the cost of duplication is minimal if not negligible, and the protection against errors in the application layer is valuable enough to negate that cost.

Reply:
@samuelhwilliams I don’t think it's that fundamental, just a trade-off which can go either way.  I think it's the experience with the sloooow (like hours) altering of the db constraint in the past which has mainly turned me off them.

In this instance though, I'm not sure we even need a character limit on things like the page title, on the database or the application. We have a human review process for catching that.

Reply:
@samuelhwilliams given we have a human-review process for publishing pages though, there's no actual harm caused by people initially having titles > 255 characters whilst it’s in draft?

The (Postgres) database was a few million rows. Quite a lot bigger than ours, but by no means massive. It's a common problem: https://stackoverflow.com/questions/5003260/expand-a-varchar-column-very-slowly-why  That said it might have been fixed in newer versions of Postgres, I've not checked.